### PR TITLE
feat(crl): add crl check support

### PR DESCRIPTION
**Pull request by thalesmg**
Fixes https://emqx.atlassian.net/browse/EMQX-9211

## PR Checklist
Please convert it to a draft if any of the following conditions are not met. Reviewers may skip over until all the items are checked:

- [X] Added tests for the changes
- [x] Changed lines covered in coverage report
- [x] Change log has been added to `changes/{ce,ee}/(feat|perf|fix)-<PR-id>.en.md` and `.zh.md` files
- [X] For internal contributor: there is a jira ticket to track this change
- [X] If there should be document changes, a PR to emqx-docs.git is sent, or a jira ticket is created to follow up
- [X] Schema changes are backward compatible

## Checklist for CI (.github/workflows) changes

- [NA] If changed package build workflow, pass [this action](https://github.com/emqx/emqx/actions/workflows/build_packages.yaml) (manual trigger)
- [NA] Change log has been added to `changes/` dir for user-facing artifacts update

### CODE REVIEW

File: apps/emqx/i18n/emqx_schema_i18n.conf [hunk 1]
```diff
@@ -1810,6 +1810,56 @@ server_ssl_opts_schema_ocsp_refresh_http_timeout {
...
+        en: "Enable CRL Check"
+        zh: "启用 CRL 检查"
+    }
+}
+
+crl_cache_refresh_http_timeout {
+    desc {
+        en: "The timeout for the HTTP request when fetching CRLs.  This is"
+            " global for all listeners."
+        zh: "获取 CRLs 时 HTTP 请求的超时。 这对所有监听器来说是全局的。"
```
Comment (zmstone) [inline apps/emqx/i18n/emqx_schema_i18n.conf:1828]:
```suggestion
        zh: "获取 CRLs 时 HTTP 请求的超时。 该配置对所有启用 CRL 检查的监听器监听器有效。"
```

File: apps/emqx/i18n/emqx_schema_i18n.conf [hunk 2]
```diff
@@ -1810,6 +1810,56 @@ server_ssl_opts_schema_ocsp_refresh_http_timeout {
...
+    label: {
+        en: "Enable CRL Check"
+        zh: "启用 CRL 检查"
+    }
+}
+
+crl_cache_refresh_http_timeout {
+    desc {
+        en: "The timeout for the HTTP request when fetching CRLs.  This is"
+            " global for all listeners."
```
Comment (zmstone) [inline apps/emqx/i18n/emqx_schema_i18n.conf:1827]:
```suggestion
        en: "The timeout for the HTTP request when fetching CRLs.  This is"
            " a global setting for all listeners."
```

File: apps/emqx/i18n/emqx_schema_i18n.conf [hunk 3]
```diff
@@ -1810,6 +1810,56 @@ server_ssl_opts_schema_ocsp_refresh_http_timeout {
...
+        en: "CRL Cache Refresh HTTP Timeout"
+        zh: "CRL 缓存刷新 HTTP 超时"
+    }
+}
+
+crl_cache_refresh_interval {
+    desc {
+        en: "The period to refresh the CRLs from the servers.  This is global"
+            " for all URLs and listeners."
+        zh: "从服务器刷新CRL的周期。 这对所有URL和监听器来说是全局性的。"
```
Comment (zmstone) [inline apps/emqx/i18n/emqx_schema_i18n.conf:1840]:
```suggestion
        en: "The period to refresh the CRLs from the servers.  This is a global setting"
            " for all URLs and listeners."
        zh: "从服务器刷新CRL的周期。 该配置对所有 URL 和监听器有效。"
```

File: apps/emqx/src/emqx_schema.erl [hunk 4]
```diff
@@ -226,6 +226,11 @@ roots(low) ->
             sc(
                 ref("trace"),
                 #{}
+            )},
+        {"crl_cache",
+            sc(
+                ref("crl_cache"),
+                #{}
```
Comment (zmstone) [inline apps/emqx/src/emqx_schema.erl:233]:
```suggestion
                #{hidden => true}
```

File: apps/emqx/src/emqx_schema.erl [hunk 5]
```diff
@@ -794,6 +799,40 @@ fields("listeners") ->
...
+    %% Note: we make the refresh interval and HTTP timeout global (not
+    %% per-listener) because multiple SSL listeners might point to the
+    %% same URL.  If they had diverging timeout options, it would be
+    %% confusing.
+    [
+        {"refresh_interval",
+            sc(
+                duration(),
+                #{
+                    hidden => true,
```
Comment (zmstone) [inline apps/emqx/src/emqx_schema.erl:812]: no need to hide here.
```suggestion
```

File: apps/emqx/src/emqx_schema.erl [hunk 6]
```diff
@@ -794,6 +799,40 @@ fields("listeners") ->
...
+                    default => <<"15m">>,
+                    desc => ?DESC("crl_cache_refresh_interval")
+                }
+            )},
+        {"http_timeout",
+            sc(
+                duration(),
+                #{
+                    hidden => true,
```
Comment (zmstone) [inline apps/emqx/src/emqx_schema.erl:821]:
```suggestion
```

File: apps/emqx/src/emqx_schema.erl [hunk 7]
```diff
@@ -794,6 +799,40 @@ fields("listeners") ->
...
+                    default => <<"15s">>,
+                    desc => ?DESC("crl_cache_refresh_http_timeout")
+                }
+            )},
+        {"capacity",
+            sc(
+                pos_integer(),
+                #{
+                    hidden => true,
```
Comment (zmstone) [inline apps/emqx/src/emqx_schema.erl:830]:
```suggestion
```

File: apps/emqx/src/emqx_schema.erl [hunk 8]
```diff
@@ -2263,13 +2317,19 @@ server_ssl_opts_schema(Defaults, IsRanchListener) ->
                             required => false,
                             validator => fun ocsp_inner_validator/1
                         }
+                    )},
+                {"crl",
+                    sc(
+                        ref("crl"),
+                        #{required => false}
```
Comment (zmstone) [inline apps/emqx/src/emqx_schema.erl:2324]:
```suggestion
                {"enable_crl_check",
                    sc(
                        boolean(),
                        #{default  => false}
```

**Review by zmstone — COMMENTED** [review]
we'll need a markdown doc which explains how crl cache works in EMQX and how it's designed.
can be added here: https://github.com/emqx/emqx-dev-docs

**Review by qzhuyan — APPROVED** [review]

**Comment by qzhuyan** [comment]
LGTM but I am not sure if dual license is ok in 
[apps/emqx/src/emqx_ssl_crl_cache.erl](https://coveralls.io/builds/58048024/source?filename=apps%2Femqx%2Fsrc%2Femqx_ssl_crl_cache.erl)

**Comment by thalesmg** [comment]
> LGTM but I am not sure if dual license is ok in [apps/emqx/src/emqx_ssl_crl_cache.erl](https://coveralls.io/builds/58048024/source?filename=apps%2Femqx%2Fsrc%2Femqx_ssl_crl_cache.erl)

I checked with @zmstone , this seems ok.  :smiley_cat:

**Pull request merged**